### PR TITLE
Fix issues with querying Maven artifacts from different repositories

### PR DESCRIPTION
**Pull request by sschuberth**
Please have a look at the individual commit messages for the details.

### CODE REVIEW

File: analyzer/src/main/kotlin/managers/utils/MavenSupport.kt [hunk 1]
```diff
@@ -644,8 +652,11 @@ class MavenSupport(private val workspaceReader: WorkspaceReader) {
      * below a "target" subdirectory of the actual project.
      */
     fun parsePackage(
-        artifact: Artifact, repositories: List<RemoteRepository>,
-        localProjects: Map<String, MavenProject> = emptyMap(), sbtMode: Boolean = false
+        artifact: Artifact,
+        repositories: List<RemoteRepository>,
+        useArtifactRepositories: Boolean,
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/utils/MavenSupport.kt:657]: Should we set this to `useArtifactRepositories = true` be default to that together with the default of `sbtMode = false` this describes the default Maven behavior?

File: analyzer/src/main/kotlin/managers/utils/MavenSupport.kt [hunk 2]
```diff
@@ -489,7 +489,11 @@ class MavenSupport(private val workspaceReader: WorkspaceReader) {
         }
     }
 
-    private fun requestRemoteArtifact(artifact: Artifact, repositories: List<RemoteRepository>): RemoteArtifact {
+    private fun requestRemoteArtifact(
+        artifact: Artifact,
+        repositories: List<RemoteRepository>,
+        useArtifactRepositories: Boolean
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/utils/MavenSupport.kt:495]: I'm open to better names here. Something like `useReposFromDependencyPoms` sounds a bit clumsy, though.

Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/utils/MavenSupport.kt:495]: @mnonnenmacher any suggestion here?

File: analyzer/src/main/kotlin/managers/utils/MavenSupport.kt (hunk 1 above)
Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/utils/MavenSupport.kt:657]: @oss-review-toolkit/kotlin-devs any opinion here?

Comment (mnonnenmacher) [reply analyzer/src/main/kotlin/managers/utils/MavenSupport.kt:657]: I have no strong opinion here, I'm fine with and without default value. But it would be very good to explain the argument in the function docs.

File: analyzer/src/main/kotlin/managers/utils/MavenSupport.kt (hunk 2 above)
Comment (mnonnenmacher) [reply analyzer/src/main/kotlin/managers/utils/MavenSupport.kt:495]: I actually like the suggestion, maybe a bit clumsy but also precise. Alternatives: `useReposFromDependency`, `useReposDeclaredByDependency`.
I assume you use plural "poms" because repositories from all POMs on the path to the dependency are used?

Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/utils/MavenSupport.kt:495]: > I assume you use plural "poms" because repositories from all POMs on the path to the dependency are used?

Exactly. So I'll probably go for `useReposFromDependencies` (plural).

**Review by mnonnenmacher — DISMISSED** [review]
Approved, but it looks like there are more expected test results you need to update.

**Comment by sschuberth** [comment]
> Approved, but it looks like there are more expected test results you need to update.

The failing test pass for me locally (also see [my message on Slack](https://ort-talk.slack.com/archives/CD1M764AJ/p1676551971669669)). Would you mind checking whether `GradleKotlinScriptFunTest` and `SbtFunTest` pass locally for you, too?

**Comment by sschuberth** [comment]
On CI, I for example see

    https://repo.maven.apache.org/maven2/org/jetbrains/annotations/13.0/annotations-13.0.jar

being used instead of

    https://jcenter.bintray.com/org/jetbrains/annotations/13.0/annotations-13.0.jar

**Comment by mnonnenmacher** [comment]
> The failing test pass for me locally (also see [my message on Slack](https://ort-talk.slack.com/archives/CD1M764AJ/p1676551971669669)). Would you mind checking whether `GradleKotlinScriptFunTest` and `SbtFunTest` pass locally for you, too?

Yes, tests pass locally.

**Comment by sschuberth** [comment]
> Yes, tests pass locally.

As I'm totally clueless by now, I've filed https://github.com/actions/setup-java/issues/454.

**Comment by mnonnenmacher** [comment]
> As I'm totally clueless by now, I've filed [actions/setup-java#454](https://github.com/actions/setup-java/issues/454).

Did you already try to clear the settings.xml after the setup-java task?

**Comment by sschuberth** [comment]
> Did you already try to clear the settings.xml after the setup-java task?

Not *after* `setup-java`, but [I instructed `setup-java` to not overwrite it](https://github.com/oss-review-toolkit/ort/pull/6489/commits/30a6dd0d24a060c61bf970eb696e1b7830037c7d) (after I created an empty one).

File: analyzer/src/main/kotlin/managers/utils/MavenSupport.kt [hunk 3]
```diff
@@ -529,7 +529,7 @@ class MavenSupport(private val workspaceReader: WorkspaceReader) {
             val localRepositories = allRepositories - remoteRepositories
             if (localRepositories.isNotEmpty()) {
                 // No need to use curly-braces-syntax for logging here as the log level check is already done above.
-                logger.debug { "Ignoring local repositories $localRepositories." }
+                logger.debug("Ignoring local repositories $localRepositories.")
```
Comment (fviernau) [inline analyzer/src/main/kotlin/managers/utils/MavenSupport.kt:532]: What's the advantage of not using curly braces?

**Review by fviernau — CHANGES_REQUESTED** [review]

File: analyzer/src/main/kotlin/managers/utils/MavenSupport.kt (hunk 3 above)
Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/utils/MavenSupport.kt:532]: Avoiding a superfluous check if the log level is set to debug.

Comment (fviernau) [reply analyzer/src/main/kotlin/managers/utils/MavenSupport.kt:532]: I believe such micro optimizations should not be valued that high, and consistency preferred in this case. Which would also allow removing that comment in line 531.

Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/utils/MavenSupport.kt:532]: > Which check actually is superflous?

The `logger.delegate.isDebugEnabled` condition duplicates the check done in the implementation if `debug {}`.

> I believe such micro optimizations should not be valued that high

I'm not valuing it high, I'm just changing the code to match the documentation.

> consistency preferred in this case.

There are also other code locations where is "micro-optimization" is performed, and I prefer to stay consistent to these.

Comment (fviernau) [reply analyzer/src/main/kotlin/managers/utils/MavenSupport.kt:532]: > I'm not valuing it high

I believe my point was clear, e.g. I believe it's inappropriate to do such kind of optimizations in the given context.

> There are also other code locations where is "micro-optimization" is performed, and I prefer to stay consistent to these.

Would you mind sharing why?

I'll share my reason as well: keep it simple.

File: analyzer/src/main/kotlin/managers/utils/MavenDependencyHandler.kt [hunk 4]
```diff
@@ -94,17 +94,21 @@ class MavenDependencyHandler(
...
+
+            // There is the corner case that a dependency references a project, but in a different version than
+            // the one used by the local build. Then, this dependency is actually a package, but Maven's
+            // resolution mechanism might prefer using the project. Therefore, the check whether the dependency
+            // is a project must be done after the package resolution again.
+            if (isLocalProject(pkg.id)) {
+                localProjectIds += dependency.identifier()
+                null
+            } else {
+                pkg
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/managers/utils/MavenDependencyHandler.kt:111]: Unrelated refactoring.

File: analyzer/src/test/kotlin/managers/utils/MavenDependencyHandlerTest.kt [hunk 5]
```diff
@@ -158,7 +158,9 @@ class MavenDependencyHandlerTest : WordSpec({
             val handler = createHandler(sbtMode = true)
 
             every { dependency.repositories } returns repos
-            every { handler.support.parsePackage(artifact, repos, sbtMode = true) } returns pkg
+            every {
+                handler.support.parsePackage(artifact, repos, sbtMode = true)
+            } returns pkg
```
Comment (mnonnenmacher) [inline analyzer/src/test/kotlin/managers/utils/MavenDependencyHandlerTest.kt:163]: Unrelated refactoring (I'd actually keep the one-liner).

File: analyzer/src/main/kotlin/managers/utils/MavenSupport.kt [hunk 6]
```diff
@@ -513,6 +508,13 @@ class MavenSupport(private val workspaceReader: WorkspaceReader) {
             repositories
         }
 
+        val cacheKey = "$artifact@$allRepositories"
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/managers/utils/MavenSupport.kt:511]: Commit message: "This, " -> "Therefore, "?

File: analyzer/src/main/kotlin/managers/utils/MavenSupport.kt (hunk 3 above)
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/managers/utils/MavenSupport.kt:532]: I'd remove the code comment instead because the performance impact is so insignificant. But I'm fine with doing this as well.

**Review by mnonnenmacher — CHANGES_REQUESTED** [review]

File: analyzer/src/main/kotlin/managers/utils/MavenSupport.kt (hunk 3 above)
Comment (mnonnenmacher) [reply analyzer/src/main/kotlin/managers/utils/MavenSupport.kt:532]: Since we know that `remoteRepositories` is a subset of `allRepositories` we could do this instead:
```kotlin
if (allRepositories.size > remoteRepositories.size) {
  logger.debug {
    val localRepositories = allRepositories - remoteRepositories
    "Ignoring local repositories $localRepositories."
  }
}
```

**Review by mnonnenmacher — CHANGES_REQUESTED** [review]

File: analyzer/src/main/kotlin/managers/utils/MavenDependencyHandler.kt (hunk 4 above)
Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/utils/MavenDependencyHandler.kt:111]: Moved to a preceding commit (there also was a left-over unneeded wrapping involved).

File: analyzer/src/test/kotlin/managers/utils/MavenDependencyHandlerTest.kt (hunk 5 above)
Comment (sschuberth) [reply analyzer/src/test/kotlin/managers/utils/MavenDependencyHandlerTest.kt:163]: Undone.

File: analyzer/src/main/kotlin/managers/utils/MavenSupport.kt (hunk 6 above)
Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/utils/MavenSupport.kt:511]: It should have been "Thus", fixed.

File: analyzer/src/main/kotlin/managers/utils/MavenSupport.kt (hunk 3 above)
Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/utils/MavenSupport.kt:532]: > I'd remove the code comment instead because the performance impact is so insignificant.

Looks like I'm outvoted here, so I'll change as proposed.

**Review by mnonnenmacher — APPROVED** [review]
